client/checkout/woopay/express-button/woopay-express-checkout-button.js

```diff
@@ -69,7 +79,141 @@ export const WoopayExpressCheckoutButton = ( {
 		}
 	}, [ isPreview, context ] );
 
-	const initWooPay = ( e ) => {
+	const newIframe = useCallback( () => {
+		if ( ! getConfig( 'isWoopayFirstPartyAuthEnabled' ) ) {
+			return;
+		}
+
+		const getWoopayOtpUrl = () => {
+			const tracksUserId = JSON.stringify(
+				getConfig( 'tracksUserIdentity' )
+			);
+
+			const urlParams = new URLSearchParams();
+			urlParams.append( 'testMode', getConfig( 'testMode' ) );
+			urlParams.append( 'source_url', window.location.href );
+			urlParams.append( 'tracksUserIdentity', tracksUserId );
+
+			return getConfig( 'woopayHost' ) + '/otp/?' + urlParams.toString();
+		};
+
+		const iframe = document.createElement( 'iframe' );
+		iframe.src = getWoopayOtpUrl();
+		iframe.height = 0;
+		iframe.style.visibility = 'hidden';
+		iframe.style.position = 'absolute';
+		iframe.style.top = '0';
+
+		iframe.addEventListener( 'load', () => {
+			// Change button's onClick handle to use express checkout flow.
+			initWoopayRef.current = ( e ) => {
+				e.preventDefault();
+
+				if ( isPreview ) {
+					return;
+				}
+
+				wcpayTracks.recordUserEvent(
+					wcpayTracks.events.WOOPAY_BUTTON_CLICK,
+					{
+						source: context,
+					}
+				);
+
+				if ( isProductPage ) {
+					const productData = getProductDataRef.current();
+
+					if ( ! productData ) {
+						return;
+					}
+
+					if ( listenForCartChanges.stop ) {
+						// Temporarily stop listening for cart changes to prevent
+						// rendering a new button + iFrame when the cart is updated.
+						listenForCartChanges.stop();
+					}
+
+					addToCartRef.current( productData ).then( () => {
+						if ( listenForCartChanges.start ) {
+							// Start listening for cart changes, again.
+							listenForCartChanges.start();
+						}
+						request(
+							buildAjaxURL(
+								getConfig( 'wcAjaxUrl' ),
+								'get_woopay_session'
+							),
+							{
+								_ajax_nonce: getConfig( 'woopaySessionNonce' ),
+							}
+						).then( ( response ) => {
+							iframe.contentWindow.postMessage(
+								{
+									action: 'setPreemptiveSessionData',
+									value: response,
+								},
+								getConfig( 'woopayHost' )
+							);
+						} );
+					} );
+				} else {
+					// Non-product pages already have pre-fetched session data.
+					sessionDataPromiseRef.current.then( ( response ) => {
```

Comment:
Perhaps we can be extra careful here and check if the promise exists before calling `then`?

```suggestion
					sessionDataPromiseRef.current?.then( ( response ) => {
```

Reply:
Although at the moment, `sessionDataPromiseRef.current` is guaranteed to be initialized, there is a chance we may refactor in the future and cause this to crash. Fixed in https://github.com/Automattic/woocommerce-payments/pull/7199/commits/530bc0908f101470adcadd9535cb71d7c4b5f662.